influxql/engine.go

```diff
@@ -1,1250 +1,554 @@
-func (r *Reducer) Reduce() *Emitter {
-	inputs := make([]<-chan map[Key]interface{}, len(r.mappers))
-	for i, m := range r.mappers {
-		inputs[i] = m.Map().C()
+func newLiteralProcessor(val interface{}) processor {
+	return func(values []interface{}) interface{} {
+		return val
 	}
-
-	e := NewEmitter(1)
-	r.c = e.C()
-	go r.run(e, inputs)
-	return e
 }
 
-func (r *Reducer) run(e *Emitter, inputs []<-chan map[Key]interface{}) {
-	// Close emitter when we're done.
-	defer func() { _ = e.Close() }()
+func getBinaryProcessor(expr *BinaryExpr, startIndex int) (processor, int) {
+	lhs, index := getProcessor(expr.LHS, startIndex)
+	rhs, index := getProcessor(expr.RHS, index)
 
-	// Buffer all the inputs.
-	bufInputs := make([]*bufInput, len(inputs))
-	for i, input := range inputs {
-		bufInputs[i] = &bufInput{c: input}
-	}
+	return newBinaryExprEvaluator(expr.Op, lhs, rhs), index
+}
 
-	// Stream data from the inputs and reduce.
-	for {
-		// Read all data from the inputers with the same timestamp.
-		timestamp := int64(0)
-		for _, bufInput := range bufInputs {
-			rec := bufInput.peek()
-			if rec == nil {
-				continue
-			}
-			if timestamp == 0 || rec.Key.Timestamp < timestamp {
-				timestamp = rec.Key.Timestamp
+func newBinaryExprEvaluator(op Token, lhs, rhs processor) processor {
+	switch op {
+	case ADD:
+		return func(values []interface{}) interface{} {
+			l := lhs(values)
+			r := rhs(values)
+			if lv, ok := l.(float64); ok {
+				if rv, ok := r.(float64); ok {
+					if rv != 0 {
+						return lv + rv
+					}
+				}
 			}
+			return nil
 		}
-
-		data := make(map[Key][]interface{})
-		for _, bufInput := range bufInputs {
-			for {
-				rec := bufInput.read()
-				if rec == nil {
-					break
-				}
-
-				if rec.Key.Timestamp != timestamp {
-					bufInput.unread(rec)
-					break
+	case SUB:
+		return func(values []interface{}) interface{} {
+			l := lhs(values)
+			r := rhs(values)
+			if lv, ok := l.(float64); ok {
+				if rv, ok := r.(float64); ok {
+					if rv != 0 {
+						return lv - rv
+					}
 				}
-
-				data[rec.Key] = append(data[rec.Key], rec.Value)
 			}
+			return nil
 		}
-
-		if len(data) == 0 {
-			break
+	case MUL:
+		return func(values []interface{}) interface{} {
+			l := lhs(values)
+			r := rhs(values)
+			if lv, ok := l.(float64); ok {
+				if rv, ok := r.(float64); ok {
+					if rv != 0 {
+						return lv * rv
+					}
+				}
+			}
+			return nil
 		}
-
-		// Sort keys.
-		keys := make(keySlice, 0, len(data))
-		for k := range data {
-			keys = append(keys, k)
+	case DIV:
+		return func(values []interface{}) interface{} {
+			l := lhs(values)
+			r := rhs(values)
+			if lv, ok := l.(float64); ok {
+				if rv, ok := r.(float64); ok {
+					if rv != 0 {
+						return lv / rv
+					}
+				}
+			}
+			return nil
 		}
-		sort.Sort(keys)
-
-		// Reduce each key.
-		for _, k := range keys {
-			r.fn(k, data[k], e)
+	default:
+		// we shouldn't get here, but give them back nils if it goes this way
```

Comment:
Quietly accepting stuff that shouldn't happen always makes me nervous. A panic might seem harsh, but if it shouldn't happen, shouldn't we find out?